_posts/2021-12-09-hhvm-4.139.markdown

```diff
@@ -0,0 +1,29 @@
+---
+title: "HHVM 4.139, and December release schedule"
+layout: post
+author: fred
+category: blog
+---
+
+HHVM 4.139 is released! This release marks the end of support for 4.133; HHVM 4.134&ndash;4.138 remain supported, as do the 4.102 and 4.128 LTS releases.
+
+# December release schedule
+
+We expect to release 4.140 next week, but there will be no 4.141 release (week of the 20th), or 4.142 (week of the 27th). We plan to resume our usual schedule with 4.143 in the week of the 3rd of January, however we consider this release 'at-risk', depending on changes during the preceeding weeks.
+
+# Highlights
+
+- Hack now supports [`readonly`](https://docs.hhvm.com/hack/readonly/introduction) data types.
+- Reliability of the VSCode debugger interface has been improved.
+- The IDE quickfix for missing methods now creates `async` stub methods if the return type is `Awaitable`.
+- Added IDE quickfix for int, float, or string constants that are declared without a type.
+- `json_decode()` is now faster, and consumes less memory.
+
+# Breaking Changes
+
+- It is now an error to read static properties in functions that do not have the `globals` or `read_globals` capabilties.
+- It is now an error to write static properties in functions that do not have the `globals` capabilty.
+- `json_decode()` is now stricter about the input it accepts (per JSON/ECMA-404):
+  - leading 0s are no longer permitted for any numbers.
```

Comment:
`01` and `-01` still parse as `1` and `-1` respectively. `[01]` and `[-01]` are now a correct syntax error 🎉.
https://github.com/facebook/hhvm/issues/8797#issuecomment-990146598 contains the test results.

Reply:
also `{"foo": 01}` is invalid

Reply:
@Matzeb, what's the intended behavior of https://github.com/facebook/hhvm/commit/d45c940a9aec162a81d2c812eb644148fa97a758 ?

```
hphpd> =json_decode('03')
3
hphpd> =json_decode('-03')
-3
hphpd> =json_decode('{"foo": 03}')
null
hphpd> =json_decode('{"foo": -03}')
null
hphpd> =json_decode('{"foo": 3}')
stdClass Object
(
    [foo] => 3
)

hphpd> =json_decode('{"foo": -3}')
stdClass Object
(
    [foo] => -3
)
```

Reply:
Don't forget to toggle the `bool $assoc` argument as well. The parser accepts more invalid inputs when passing the non-default value `true`.

Reply:
@lexidor are you seeing any observable behavior change with leading 0s? Comparing to 4.135, I'm not seeing one

Reply:
Yep, `json_decode('{"foo": -03}', true)` is one

Reply:
I was comparing against 4.88. That was when I wrote the original issue.